man/ffpcomplex.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpcomplex 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpcomplex - This program filters FFP profiles based on the complexity of features.
.SH SYNOPSIS
.BI "ffpcomplex [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
This utility can be used to filter high or low complexity features
from the FFP matrix. The upper and/or lower limits can be set 
using a complexity cuttoff (the default) or a probability cutoff can be used
assuming a normal  distribution.  Probability limits are
the cummulative probability values from the normal distribution
found from the calculated mean and stdev of the input FFP. 
Requires key-valued FFP input.
FFPs will be read from standard input if no options are supplied and 
.B ffpcomplex  
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list. The default input type is assumed to be nucleotide sequence.
.SH OPTIONS
.TP
.B -s, --stats
.PP
Print statistics describing the FFP complexity distribution to
standard error.
.TP
.BI "-u " "FLOAT, " "--upper=" "FLOAT"
.PP
.RI "The upper threshold limit for feature complexity where " "FLOAT" " can be "
.RI "any complexity value or for probability distributions, in which case" " FLOAT "
can range from 0 to 1 -- representing the probability in a cummulative distribution
function.
.RB "This can be used in conjunction with " "-l" " or by itself."
.TP
.BI "-l " "FLOAT, " "--lower=" "FLOAT"
.PP
.RB "The lower threshold limit. This can be used in conjuntion with " "-u" " or by itself."
.TP
.B -n, --norm
.PP
Use a normal distribution for filtering. Default is the raw complexity values.
.TP 
.B -d, --disable
.PP
Disable classing of amino acid and nucleotide sequences. 
.TP
.B -a, --amino
.PP
Input FFPs are animo acid sequences.
.TP
.B -t, --text
.PP
Input FFPs are text.
.TP
.B -h, --help
.PP
Display short help message.
.TP
.B -v, --version
.PP
Display version information.
.PP
.SH EXAMPLES
The main usage of this progam is for filtering out low complexity
features, for example features which primarily represent repetitive
DNA might be need to be filtered out -- since they don't provide much
useful phylogenetic signal. To get an idea about what the complexity
of the words in your FFPs are you can use the 
.B \-s 
option to get a statistical report, which includes the mean, range
and standard deviation of complexity.  The complexity of a feature
is determined by finding the total entropy of all individual sub features
for all 
.I k
less than 
.I l,
which is the length of the feature.  For example given the feature
.PP
.CODE GCGCGCGC
.PP 
determine the entropy of 
.IR "k" "=1,"
which which is H(1) = f(G)log f(G) + f(C)log f(C).  Next determine the entropy
of all 
.iR "k" "=2," 
sub-mers, H(2) = f(GC)log f(GC) + f(CG)log f(CG).  Repeat for all 
.I k 
less than
.IR "l" "."   
The total complexity of the 
.IR "l" "-mer" 
is H(l) = H(1) + H(2) + ... + H(l-1).
Here we filter out all features with complexity less than 1.0:
.PP
.CODE ffpry -l 5 *.fna > vectors
.CODE ffpcomplex -l 1.0 vectors > vectors.filt
.PP
Features with high and low complexities can 
be removed simultaneously.  This example excludes all features which
with complexity greater than 10 and less than 2.
.PP
.CODE ffpcomplex -u 10.0 -l 2.0
.PP
Finally we can assume that the complexities of the features used follow a specific
distribution (usually the normal distribution is the appropriate choice).
We can exclude features that lie outside a certain range of the cummulative
probability distribution.  Here we exclude all features which lie below
10% and above 95% of the normal CDF.
.PP
.CODE ffpcomplex -n -l 0.1 -u 0.95
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpry(1),
.BR ffpfilt(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpcomplex.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <float.h>
#include <math.h>
#include <string.h>
#include <getopt.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hash.h"
#include "utils.h"
#include "vstring.h"
#include "cdfmacros.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[FILENAME_MAX];
#define MAX_WORD_LENGTH 40 /**< Maximum feature length */

void hashCol(FILE * fp);
int isKeyBased(FILE * fp);
int getKeyLength(FILE * fp);
float complexity(char *key);

char usage_str[] = "Usage: %s [OPTIONS]... [FILE]...\n\
This program filters words by complexity\n\n\
Given no options, the default behavior of the program is to\n\
assume Nucleic acids FFP input\n\
\t-a, --amino\tInput is amino acids\n\
\t-t, --text\tInput is text\n\
\t-d, --disable\tDisable classing of AAs and Nuc.\n\
\t-n, --normal\tAssume normal CDF distribution of complexity.\n\
\t-l, --lower\tLower limit for filtration, with -n this is a\n\
\t\t\tprobability.\n\
\t-u, --upper\tUpper limit for filtration, with -n this is a\n\
\t\t\tprobability.\n\
\t-s, --stats\tPrint complexity stats to stderr\n\
\t-h, --help\tThis text.\n\
\t-v, --version\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";

char *weightVector;
int Length = MAX_WORD_LENGTH;
			    /**< Feature length */

bool flagK = false;
bool flagA = false;
bool flagT = false;
bool flagD = false;
bool flagF = true;
bool flagU = false;
bool flagL = false;
bool flagN = false;
bool flagS = false;

float lower = -FLT_MAX;
float upper = FLT_MAX;

int main(int argc, char **argv)
{
    FILE *fp = NULL;
    struct stat input;
    int opt;

    int mode = nucleotide;
    int option_index = 0;

    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{"disable", no_argument, 0, 'd'},
	{"lower", required_argument, 0, 'l'},
	{"upper", required_argument, 0, 'u'},
	{"version", no_argument, 0, 'v'},
	{"normal", no_argument, 0, 'n'},
	{"stats", no_argument, 0, 's'},
	{0, 0, 0, 0}
    };


    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "hatdvu:l:ns",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 'a':
	    flagA = !flagA;
	    break;
	case 't':
	    flagT = !flagT;
	    break;
	case 'd':
	    flagD = !flagD;
	    break;
	case 'n':
	    flagN = !flagN;
	    flagF = !flagF;
	    break;
	case 'l':
	    flagL = !flagL;
	    lower = atof(optarg);
	    break;
	case 'u':
	    flagU = !flagU;
	    upper = atof(optarg);
	    break;
	case 's':
	    flagS = !flagS;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    break;
	}


    if (flagA && flagT)
	fatal_msg("Option -a or -t not both\n");


    if (flagA)
	mode = amino;
    else if (flagT)
	mode = text;

    initHash(0, mode, !flagD);

    if ((argc - optind) > 1)
	fatal_msg("Specify only one File argument\n");

    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
		fatal_msg("%s: %s.\n", *argv,strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

    	// process if its a pipe

   	 if (!isRegularFile(fp)) {
		fp = convertPipeToFile(fp);
    	}
    	// gather complexity stats.

	if (!isKeyBased(fp))
	   fatal_msg("%s: Not a key valued FFP.\n", *argv);

	Length = getKeyLength(fp);
	hashCol(fp);

	fclose(fp);

    } while (*argv);

    return EXIT_SUCCESS;
}

/**
 * Calculate the complexity of a feature, with length n.
 * The frequencies are calcualted for all k-mers from
 * length 1 to n-1 which are sub-words of the feature. 
 * From this the log2 based Shannon Entropy is calculated.  
 * Lower values indicate lower complexity.
 *
 * @param key an amino acid, nucleotide or text feature 
 * @return float Value indicates the Shannon entropy 
 * @todo It would simplify the use of complexity for filtration if this function had its own
 *       Internal hash variable
 */


float complexity(char *key)
{
    int L, i, N;
    char s[Length];
    float entropy = 0;
    char **keys;
    N = 0;
    L = 1;

    while (L < Length) {
	for (i = 0; i < Length - L + 1; i++) {
	    strncpy(s, &key[i], L);
	    s[L] = '\0';
	    hashInc(s);
	}
	L++;
	N += L;
    }

    hashKeys(&keys);

    for (i = 0; i < numKeys(); i++)
	entropy -=
	    (float) hashval(keys[i]) / N * log2((float) hashval(keys[i]) / N);

    freeHash();

    return entropy;
}





/**
 * Convert a (key,value) FFP to a columnar FFP
 *
 * FFP in fp must be columnar FFP.
 *
 * @param fp A file pointer to a (key,value) FFP
 * @return none
 */



void hashCol(FILE * fp)
{
    unsigned val;
    char s[Length];
    char c[2];
    bool deleteKey;
    float cval;
    float cvalMin=FLT_MAX;
    float cvalMax=0;
    double x1 = 0;
    double x2 = 0;
    double sigma;
    double zscore;
    unsigned ctr = 0;
    unsigned items;
    unsigned lineno=0;
    unsigned line_offset=0;

    errno=0;
    while ( (items=fscanf(fp, "%s %u%[\r\n]", s, &val,c)) != EOF ) { 
	//fprintf(stderr,"%s %d\n",strerror(errno),items);
	if (errno) 
		fatal_msg("Parse error at line %u char %ld: %s.\n",
			lineno,ftell(fp)-line_offset, strerror(errno) );
	
	switch (items) {
	        case 3:
			line_offset=ftell(fp);
			lineno++;	  
		case 2:
			cval = complexity(s);
			x1 += cval;
			x2 += cval * cval;
			ctr++;
		break;	
		default:
			fatal_msg("Parse error at line %u char %ld.\n",
			lineno,ftell(fp)-line_offset);
		break;
	}


	if (flagS) {
		if (cval < cvalMin) 
			cvalMin=cval;
		if (cval > cvalMax)
			cvalMax=cval;
	}

    }


    x1 /= ctr;
    x2 /= ctr;

    sigma = sqrt(x2 - x1 * x1);

    if (flagS){
	fprintf(stderr, "Avg\tstddev\tmin\tmax\n");
	fprintf(stderr, "%lf\t%lf\t%f\t%f\n", x1, sigma,cvalMin,cvalMax);
    }

    rewind(fp);

    // perform tests here to decide whether to keep ffp.

    errno=0;
    while ( (items=fscanf(fp, "%s %u%[\r\n]", s, &val,c)) != EOF ) { 
	//fprintf(stderr,"%s %d\n",strerror(errno),items);
	if (errno) 
		fatal_msg("Parse error at line %u char %ld: %s.\n",
			lineno,ftell(fp)-line_offset, strerror(errno) );
	
	    cval = complexity(s);
	    zscore = ((float) cval - x1) / sigma;

	    deleteKey = false;

	    if (flagU) {
		if (flagF && cval > upper)
		    deleteKey = true;
		if (flagN && norm_cdf(zscore) > upper)
		    deleteKey = true;
	    }

	    if (flagL) {
		if (flagF && cval < lower)
		    deleteKey = true;
		if (flagN && norm_cdf(zscore) < lower)
		    deleteKey = true;
	    }

	    if (!deleteKey)
		printf("%s %u", s, val);


	switch (items) {
	        case 3:
			printf("\n");
			line_offset=ftell(fp);
			lineno++;
	     	break;		
		case 2:
			printf("\t");
		break;	
		default:
			fatal_msg("Parse error at line %u char %ld.\n",
			lineno,ftell(fp)-line_offset);
		break;
	}
    }

}
